File: packages/assistant-protocol/src/compatibility.ts
import { SOFTWARE_VERSION, MIN_COMPATIBLE_VERSION } from "./generated/assistant-protocol.ts";
import type { ClientCompatibility, RuntimeCompatibilityError, RuntimeCompatibilityErrorCode } from "./generated/assistant-protocol.ts";

/** 不读取 Host 响应来构造页面版本；每个产品消费自己的构建常量。 */
export function currentCompatibility(): ClientCompatibility {
  return { version: SOFTWARE_VERSION, min_compatible_version: MIN_COMPATIBLE_VERSION };
}

export function compatibilityHeaders(): Record<string, string> {
  return { "x-ez-client-version": SOFTWARE_VERSION, "x-ez-min-compatible-version": MIN_COMPATIBLE_VERSION };
}

function parse(value: unknown): number[] | null {
  if (typeof value !== "string" || value.trim() !== value || !/^(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)$/.test(value)) return null;
  const parts = value.split(".").map(Number);
  return parts.some((part) => part > 0xffffffff) ? null : parts;
}

function compare(left: number[], right: number[]): number {
  return left.map((value, index) => value - right[index]).find((value) => value !== 0) ?? 0;
}

function declaration(value: unknown): ClientCompatibility | null {
  if (!value || typeof value !== "object") return null;
  const candidate = value as Record<string, unknown>;
  const version = parse(candidate.version), minimum = parse(candidate.min_compatible_version);
  if (!version || !minimum || compare(minimum, version) > 0) return null;
  return { version: candidate.version as string, min_compatible_version: candidate.min_compatible_version as string };
}

/** 与 Rust 同源向量测试；未知数据必须先校验，错误中只保留有效版本对。 */
export function checkCompatibility(client: unknown, host: unknown): RuntimeCompatibilityError | null {
  const c = declaration(client), h = declaration(host);
  let code: RuntimeCompatibilityErrorCode;
  if (!h) code = "invalid_declaration";
  else if (client == null) code = "missing_declaration";
  else if (!c) code = "invalid_declaration";
  else if (compare(parse(c.version)!, parse(h.min_compatible_version)!) < 0) code = "client_too_old";
  else if (compare(parse(h.version)!, parse(c.min_compatible_version)!) < 0) code = "host_too_old";
  else return null;
  return { code, client: c, host: h };
}

File: crates/assistant-protocol/src/compatibility.rs
//! 应用协议按双方软件版本与最低兼容版本准入；不依赖传输、存储或客户端实现。

use serde::{Deserialize, Serialize};
use ts_rs::TS;

use crate::{MIN_COMPATIBLE_VERSION, SOFTWARE_VERSION, parse_software_version};

pub const CLIENT_VERSION_HEADER: &str = "x-ez-client-version";
pub const MIN_COMPATIBLE_VERSION_HEADER: &str = "x-ez-min-compatible-version";

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[ts(export_to = "assistant-protocol.ts")]
pub struct ClientCompatibility {
    pub version: String,
    pub min_compatible_version: String,
}

impl ClientCompatibility {
    pub fn current() -> Self {
        Self {
            version: SOFTWARE_VERSION.into(),
            min_compatible_version: MIN_COMPATIBLE_VERSION.into(),
        }
    }

    pub fn is_valid(&self) -> bool {
        matches!((parse_software_version(&self.version), parse_software_version(&self.min_compatible_version)),
            (Some(version), Some(minimum)) if minimum <= version)
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(rename_all = "snake_case")]
#[ts(export_to = "assistant-protocol.ts")]
pub enum RuntimeCompatibilityErrorCode {
    MissingDeclaration,
    InvalidDeclaration,
    ClientTooOld,
    HostTooOld,
}

/// 仅回传已经验证的版本对，非法原始文本不能进入诊断。
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[ts(export_to = "assistant-protocol.ts")]
pub struct RuntimeCompatibilityError {
    pub code: RuntimeCompatibilityErrorCode,
    pub client: Option<ClientCompatibility>,
    pub host: Option<ClientCompatibility>,
}

pub fn check_compatibility(
    client: Option<&ClientCompatibility>,
    host: &ClientCompatibility,
) -> Result<(), RuntimeCompatibilityError> {
    use RuntimeCompatibilityErrorCode::*;
    let code = match client {
        _ if !host.is_valid() => InvalidDeclaration,
        None => MissingDeclaration,
        Some(client) if !client.is_valid() => InvalidDeclaration,
        Some(client)
            if parse_software_version(&client.version)
                < parse_software_version(&host.min_compatible_version) =>
        {
            ClientTooOld
        }
        Some(client)
            if parse_software_version(&host.version)
                < parse_software_version(&client.min_compatible_version) =>
        {
            HostTooOld
        }
        Some(_) => return Ok(()),
    };
    Err(RuntimeCompatibilityError {
        code,
        client: client.filter(|value| value.is_valid()).cloned(),
        host: host.is_valid().then(|| host.clone()),
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[derive(Deserialize)]
    struct Case {
        name: String,
        client: Option<ClientCompatibility>,
        host: ClientCompatibility,
        error: Option<RuntimeCompatibilityErrorCode>,
    }

    #[test]
    fn shared_software_compatibility_vectors() {
        let cases: Vec<Case> = serde_json::from_str(include_str!(
            "../../../packages/assistant-protocol/fixtures/compatibility.json"
        ))
        .unwrap();
        for case in cases {
            let outcome = check_compatibility(case.client.as_ref(), &case.host);
            assert_eq!(
                outcome.as_ref().err().map(|error| error.code),
                case.error,
                "{}",
                case.name
            );
            if let Err(error) = outcome {
                assert!(
                    error
                        .client
                        .as_ref()
                        .is_none_or(ClientCompatibility::is_valid)
                );
                assert!(
                    error
                        .host
                        .as_ref()
                        .is_none_or(ClientCompatibility::is_valid)
                );
            }
        }
    }
}
